plugins/inputs/modbus/modbus.go

```diff
@@ -173,552 +167,253 @@ func (m *Modbus) Init() error {
 	}
 
-	m.registers = append(m.registers, register{name, registersRange, fields})
-
 	return nil
 }
 
-// Connect to a MODBUS Slave device via Modbus/[TCP|RTU|ASCII]
-func connect(m *Modbus) error {
+func (m *Modbus) initClient() error {
 	u, err := url.Parse(m.Controller)
 	if err != nil {
 		return err
 	}
 
 	switch u.Scheme {
 	case "tcp":
-		var host, port string
-		host, port, err = net.SplitHostPort(u.Host)
-		if err != nil {
-			return err
-		}
-		m.tcpHandler = mb.NewTCPClientHandler(host + ":" + port)
-		m.tcpHandler.Timeout = time.Duration(m.Timeout)
-		m.tcpHandler.SlaveId = byte(m.SlaveID)
-		m.client = mb.NewClient(m.tcpHandler)
-		err := m.tcpHandler.Connect()
+		host, port, err := net.SplitHostPort(u.Host)
 		if err != nil {
 			return err
 		}
-		m.isConnected = true
-		return nil
+		handler := mb.NewTCPClientHandler(host + ":" + port)
+		handler.Timeout = time.Duration(m.Timeout)
+		m.handler = handler
 	case "file":
-		if m.TransmissionMode == "RTU" {
-			m.rtuHandler = mb.NewRTUClientHandler(u.Path)
-			m.rtuHandler.Timeout = time.Duration(m.Timeout)
-			m.rtuHandler.SlaveId = byte(m.SlaveID)
-			m.rtuHandler.BaudRate = m.BaudRate
-			m.rtuHandler.DataBits = m.DataBits
-			m.rtuHandler.Parity = m.Parity
-			m.rtuHandler.StopBits = m.StopBits
-			m.client = mb.NewClient(m.rtuHandler)
-			err := m.rtuHandler.Connect()
-			if err != nil {
-				return err
-			}
-			m.isConnected = true
-			return nil
-		} else if m.TransmissionMode == "ASCII" {
-			m.asciiHandler = mb.NewASCIIClientHandler(u.Path)
-			m.asciiHandler.Timeout = time.Duration(m.Timeout)
-			m.asciiHandler.SlaveId = byte(m.SlaveID)
-			m.asciiHandler.BaudRate = m.BaudRate
-			m.asciiHandler.DataBits = m.DataBits
-			m.asciiHandler.Parity = m.Parity
-			m.asciiHandler.StopBits = m.StopBits
-			m.client = mb.NewClient(m.asciiHandler)
-			err := m.asciiHandler.Connect()
-			if err != nil {
-				return err
-			}
-			m.isConnected = true
-			return nil
-		} else {
+		switch m.TransmissionMode {
+		case "RTU":
+			handler := mb.NewRTUClientHandler(u.Path)
+			handler.Timeout = time.Duration(m.Timeout)
+			handler.BaudRate = m.BaudRate
+			handler.DataBits = m.DataBits
+			handler.Parity = m.Parity
+			handler.StopBits = m.StopBits
+			m.handler = handler
+		case "ASCII":
+			handler := mb.NewASCIIClientHandler(u.Path)
+			handler.Timeout = time.Duration(m.Timeout)
+			handler.BaudRate = m.BaudRate
+			handler.DataBits = m.DataBits
+			handler.Parity = m.Parity
+			handler.StopBits = m.StopBits
+			m.handler = handler
+		default:
 			return fmt.Errorf("invalid protocol '%s' - '%s' ", u.Scheme, m.TransmissionMode)
 		}
 	default:
-		return fmt.Errorf("invalid controller")
+		return fmt.Errorf("invalid controller %q", m.Controller)
 	}
-}
 
-func disconnect(m *Modbus) error {
-	u, err := url.Parse(m.Controller)
-	if err != nil {
-		return err
-	}
+	m.handler.SetSlave(m.SlaveID)
+	m.client = mb.NewClient(m.handler)
+	m.isConnected = false
```

Comment:
didn't the previous version try to connect here?

Reply:
True, however modbus devices are usually quite restricted in the number of open connections, so I removed the connect here as we always connect in the gather cycle if not already connected. The idea was to connect as late as possible.

One feature that is requested (and will be committed on top of this PR) is to connect/disconnect in each cycle to allow other collection agents to access the device. This is especially useful for serial connections...